### PR TITLE
Delay load pages

### PR DESCRIPTION
**Pull request by kicken**
Delays scanning the content directory for pages until the pages variable is used.  Speeds  up loading time if the variable is not required.

### CODE REVIEW

**Comment by NeoBlack** [comment]
nice approach for #256 
this works for twig only, maybe we could implement it on core level. So PagesCollection should be located in lib/Phile/Core

File: lib/Phile/Repository/Page.php [hunk 1]
```diff
@@ -171,7 +119,7 @@ public function getPageOffset(\Phile\Model\Page $page, $offset = 0) {
 	 *
 	 * @return mixed|\Phile\Model\Page
 	 */
-	protected function getPage($filePath, $folder = CONTENT_DIR) {
+	public function getPage($filePath, $folder = CONTENT_DIR) {
```
Comment (Schlaefer) [inline lib/Phile/Repository/Page.php]: We should try to keep that protected. Users are supposed to call `findByPath`.

**Comment by Schlaefer** [comment]
I'm :+1: on the idea and in principal I'm OKish with the implementation. But of course I'm having a small "but" and here it comes. :wink:

What's a little bit uncomfortable is to split the worker code of `findByPath` and `findAll` – both involving "loading from filesystem" – into different source files. They should stay together or at least imho not split up just for implementing lazy-loading `findAll`. Imho.

I suggest we use a proxy-pattern and subclass PageCollection from Repository\Page and keep `findByPath` and `findAll` in Repository\Page. Along those lines:

In Repository\Page:

```php
class Page {

	public function findAll(array $options = array(), $folder = CONTENT_DIR) 	{
    	     return new PageProxy($options, $folder);
	}

	protected function _findAll(array $options = array(), $folder) {
		// ... existing code
	}
```

A $folder setter in the constructor needs to be implemented, but imho that's not bad thing to have anyway.

Then use the simple PageCollection from the PR first commit just as proxy extending Page:

```php
class PageProxy extends Page implements \ArrayAccess, \IteratorAggregate {

    // ...

    private function load() {
        if ($this->pages === null) {
          $this->pages = parent::_findAll($this->options, $this->folder);
        }
   }
```

**Comment by kicken** [comment]
I was not thrilled with the code separation initially either. I will give the proxy method a shot, that sounds like a good approach.

**Comment by Schlaefer** [comment]
I think we can run with that. :+1:

**Comment by Schlaefer** [comment]
merged into develop in 62496b5f2b8a2b4a29a68efa7251fd3888e125b1

**Pull request closed without merging**